include/mbedtls/fsio.h

```diff
@@ -0,0 +1,244 @@
+/**
+ * \file fsio.h
+ *
+ * \brief File read/write functions
+ *
+ *  Copyright (C) 2017, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+#ifndef MBEDTLS_FILE_H
+#define MBEDTLS_FILE_H
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#include <stddef.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if defined(MBEDTLS_FS_IO)
+/* If MBEDTLS_FS_IO is enabled then file IO functions should be made available
+ * via standard library or platform specific implementation. */
+
+/* Default and alternate implementation specific interfaces. */
+#if !defined(MBEDTLS_FS_IO_ALT)
+#include <stdio.h>
+#include <dirent.h>
+
+#define mbedtls_file_t      FILE
+#define mbedtls_fread       fread
+#define mbedtls_fgets       fgets
+#define mbedtls_fwrite      fwrite
+#define mbedtls_fclose      fclose
+#define mbedtls_ferror      ferror
+#define mbedtls_fseek       fseek
+#define mbedtls_ftell       ftell
+#define MBEDTLS_SEEK_SET    SEEK_SET
+#define MBEDTLS_SEEK_CUR    SEEK_CUR
+#define MBEDTLS_SEEK_END    SEEK_END
+
+#define mbedtls_dir_t           DIR
+#define mbedtls_opendir         opendir
+#define mbedtls_closedir        closedir
+#define MBEDTLS_FSIO_DT_BLK     DT_BLK
+#define MBEDTLS_FSIO_DT_CHR     DT_CHR
+#define MBEDTLS_FSIO_DT_DIR     DT_DIR
+#define MBEDTLS_FSIO_DT_FIFO    DT_FIFO
+#define MBEDTLS_FSIO_DT_LNK     DT_LNK
+#define MBEDTLS_FSIO_DT_REG     DT_REG
+#define MBEDTLS_FSIO_DT_SOCK    DT_SOCK
+#define MBEDTLS_FSIO_DT_UNKNOWN DT_UNKNOWN
+
+#else /* !MBEDTLS_FS_IO_ALT */
+
+/**
+ * file context.
+ *
+ */
+typedef void mbedtls_file_t;
+
```

Comment:
I'd prefer to not force the use of a pointer type for file handles. While for the POSIX file system interface, that's fine, it is not if one e.g. wants to directly use the Unix system calls, which use `int` as the file descriptor type. Forcing pointer types mandates the use of conversions between pointers and integers, such as `INT_TO_FILE_PTR` in `fsio.c`, which seems fragile.

Reply:
Changed this to ```typedef int mbedtls_file_t;``` and related code.